go-controller/pkg/observability/observability.go

```diff
@@ -0,0 +1,294 @@
+	// getCollectorKey() => collector.UUID
+	dbCollectors map[string]string
+	// cleaning up unused collectors may take time and multiple retries, as all referencing samples must be removed first.
+	// Therefore, we need to save state between those retries.
+	// getCollectorKey() => collector.SetID
+	unusedCollectors              map[string]int
+	unusedCollectorsRetryInterval time.Duration
+	// Only maxCollectorID collectors are allowed, each should have unique ID.
+	// this set is tracking already assigned IDs.
+	takenCollectorIDs sets.Set[int]
+}
+
+func NewManager(nbClient libovsdbclient.Client) *Manager {
+	return &Manager{
+		nbClient:                      nbClient,
+		collectorsLock:                sync.Mutex{},
+		dbCollectors:                  make(map[string]string),
+		unusedCollectors:              make(map[string]int),
+		unusedCollectorsRetryInterval: time.Minute,
+		takenCollectorIDs:             sets.New[int](),
+	}
+}
+
+func (m *Manager) SamplingConfig() *libovsdbops.SamplingConfig {
+	return m.sampConfig
+}
+
+func (m *Manager) Init() error {
+	// this will be read from the kube-api in the future
+	currentConfig := &collectorConfig{
+		collectorSetID: DefaultObservabilityCollectorSetID,
+		featuresProbability: map[libovsdbops.SampleFeature]int{
+			libovsdbops.EgressFirewallSample:     100,
+			libovsdbops.NetworkPolicySample:      100,
+			libovsdbops.AdminNetworkPolicySample: 100,
+			libovsdbops.MulticastSample:          100,
+			libovsdbops.UDNIsolationSample:       100,
+		},
+	}
+
+	return m.initWithConfig(currentConfig)
+}
+
+func (m *Manager) initWithConfig(config *collectorConfig) error {
+	if err := m.setSamplingAppIDs(); err != nil {
+		return err
+	}
+	if err := m.setDbCollectors(); err != nil {
+		return err
+	}
+
+	featuresConfig, err := m.addCollector(config)
+	if err != nil {
+		return err
+	}
+	m.sampConfig = libovsdbops.NewSamplingConfig(featuresConfig)
+
+	// now cleanup stale collectors
+	m.deleteStaleCollectorsWithRetry()
+	return nil
+}
+
+func (m *Manager) setDbCollectors() error {
+	m.collectorsLock.Lock()
+	defer m.collectorsLock.Unlock()
+	clear(m.dbCollectors)
+	collectors, err := libovsdbops.ListSampleCollectors(m.nbClient)
+	if err != nil {
+		return fmt.Errorf("error getting sample collectors: %w", err)
+	}
+	for _, collector := range collectors {
+		collectorKey := getCollectorKey(collector.SetID, collector.Probability)
+		m.dbCollectors[collectorKey] = collector.UUID
+		m.takenCollectorIDs.Insert(collector.ID)
+		// all collectors are unused, until we update existing configs
+		m.unusedCollectors[collectorKey] = collector.ID
+	}
+	return nil
+}
+
+// Stale collectors can't be deleted until all referencing Samples are deleted.
+// Samples will be deleted asynchronously by different controllers on their init with the new Manager.
+// deleteStaleCollectorsWithRetry will retry, considering deletion should eventually succeed when all controllers
```

Comment:
What if deletion never succeeds (e.g., due to a bug somewhere)?  Can we add a hard retry limit?  Would that be useful?

Reply:
Considering this will retry every minute, that probably is not too bad to retry indefinitely, but I can add retry limit.
I have seen very long initial syncs up to 20 min for sure (that should be completed before this cleanup can succeed), so setting retry limit to 1 hour to be safe.